### PR TITLE
Change SocketInterceptor

### PR DESCRIPTION
**Pull request by yemreinci**
Depends on #629.

Changes SocketInterceptor to be a concrete type, user passes an handler function instead of inheriting.

### CODE REVIEW

File: hazelcast/include/hazelcast/client/SocketInterceptor.h [hunk 1]
```diff
@@ -15,25 +15,57 @@
...
+            }
+
+
+        private:
+            friend class connection::ClientConnectionManagerImpl;
+
+            using handler_t = std::function<void(const Socket &)>;
 
-            virtual ~SocketInterceptor();
+            handler_t connect{ util::noop<const Socket &> };
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/SocketInterceptor.h:68]: private member naming `connect` -> `connect_`. Seems like we also forget this for other listeners as well. It is not very critical.

**Review by ihsandemir — APPROVED** [review]

File: hazelcast/include/hazelcast/client/SocketInterceptor.h (hunk 1 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/SocketInterceptor.h:68]: Ignore that for now, I've been able to automatically refactor all private data member names to in `lower_case_` format using clang-tidy: https://github.com/hazelcast/hazelcast-cpp-client/pull/632. They'll be fixed once we merge that.

**Comment by yemreinci** [comment]
verify-linux

**Comment by yemreinci** [comment]
verify-linux

**Comment by yemreinci** [comment]
verify-linux

**Comment by yemreinci** [comment]
verify-linux

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**